rfcs/008-map-indexes.md

```diff
@@ -0,0 +1,159 @@
+
+{?DATABASE, ?VIEWS, <view_signature>, ?VIEWS, ?ID_INDEX, <_id>, <view_id>} -> [emitted keys]
+
+Each index will be built and updated via the Background job queue [RFC Link TBD]. When a request for a view is received, the request process will add a job item onto the background queue for the index to be updated. A worker will take the item off the queue and update the index. Once the index has been built, the request will return with the results. This process can also be optimised in two ways. Firstly, using a new couch_events system to listen for document changes in an database and then adding indexing jobs to the queue to keep indexes warm. The second optimisation is if the index only requires a small update, rather update the index in the http request process instead of doing the work via the background queue.
+
+Initially the building of an index will be a single worker running through the changes feed and creating the index. Ideally it would be nice to parallelise that work so that multiple workers could build the index at the same time. This will reduce build times. This can be done by fetching the boundary keys for the changes feed, splitting those key ranges amongst different workers to build different parts of the index. This will require that for each document update processed, the worker must check the revision key space to determine if the document is the latest revision of the document. If it is not it should discard it. The other requirement is that CouchDB could only start serving the index once the update sequence is up to date AND all the workers have completed building their section of the index.
+
+### View clean up
+
+When a design document is changed, new indexes will be built and grouped under a new  `View Signature` . The old map indexes will still be in FDB.  CouchDB will need to run a clean up process that scans all the `View Signature`s and links them to a design document. If a `View Signature` is not linked to any design doc, that `View Signature` namespace can be removed. This will be done via the background jobs scheduler.
+
+### Stale = “ok” and stable = true
+
+ With the consistency guarantee’s CouchDB will get from FDB,  stable = true will no longer be an option that CouchDB would support. Similar `stale = “ok”` would now be translated to `update = false`
+
+### Size limits
+
+* Emitted keys will not be able to exceed 10 KB
+* Values cannot exceed 100 KB
+* There could be rare cases where the number of key-value pairs emitted for a map function could lead to a transaction either exceeding 10 MB in size which isn’t allowed or exceeding 5 MB which impacts the performance of the cluster. Ideally CouchDB will need to detect these situations and split the transaction into smaller transactions
```

Comment:
Given that these are the underlying fdb limits we may want to consider dropping the actual exposed limits down. Or else we'll end up with customers complaining that 10KiB - 1byte doesn't work because we reserve a variable number of bytes for keys and I could see similar things if we end up structuring keys.

Also, as we've seen before, actually counting how large a value is depends on a number of things which makes exact counts a bit fuzzy.

Reply:
> Ideally CouchDB will need to detect these situations and split the transaction into smaller transactions

Just a heads up that going this route would be super duper complicated make work correctly. The update_seq tracking would end up getting a lot more complicated as each individual committed transaction will have to record its progress through what was originally designed to be a single transaction.

This also brings up the parallel view building mentioned (while not proposed for V1) would also require some what to ensure that all of the parallel workers are recording their progress in such a way that they could be properly restarted in such a way that would not leave an inconsistent index.

Reply:
Good points. I've added a bit more information around those limit. I think we need to do a little experimentation before defining what a user's limit should actually be.